src/core.js

```diff
@@ -225,28 +229,24 @@ jQuery.extend( {
 	},
 
 	isPlainObject: function( obj ) {
-		var key;
+		var proto, Ctor;
 
-		// Not plain objects:
-		// - Any object or value whose internal [[Class]] property is not "[object Object]"
-		// - DOM nodes
-		// - window
-		if ( jQuery.type( obj ) !== "object" || obj.nodeType || jQuery.isWindow( obj ) ) {
+		// Detect obvious negatives
+		// Use toString instead of jQuery.type to catch host objects
+		if ( !obj || toString.call( obj ) !== "[object Object]" ) {
 			return false;
 		}
 
-		// Not own constructor property must be Object
-		if ( obj.constructor &&
-				!hasOwn.call( obj, "constructor" ) &&
-				!hasOwn.call( obj.constructor.prototype || {}, "isPrototypeOf" ) ) {
-			return false;
-		}
+		proto = getProto( obj );
 
-		// Own properties are enumerated firstly, so to speed up,
-		// if last one is own, then all properties are own
-		for ( key in obj ) {}
+		// Objects with no prototype (e.g., `Object.create( null )`) are plain
+		if ( !proto ) {
+			return true;
+		}
 
-		return key === undefined || hasOwn.call( obj, key );
+		// Objects with prototype are plain iff they were constructed by a global Object function
+		Ctor = hasOwn.call( proto, "constructor" ) && proto.constructor;
+		return typeof Ctor === "function" && fnToString.call( Ctor ) === ObjectFunctionString;
```

Comment:
With the caveat that variation that variation is always a risk, this logic works on all known environments (including weird ones like Rhino) and will be even more solid if https://tc39.github.io/Function-prototype-toString-revision/ is accepted. And if this logic fails, or any environment _does_ change, we'll get a bug report and can react accordingly.

Note also that the current logic isn't just inefficient, it is incomplete, and fails on empty-object prototypes (PR updated). And as for the cross-realm concern, that's [already tested](https://github.com/gibson042/jquery/blob/2986/test/unit/core.js#L359-L364), but could be expanded if you like.